activesupport/lib/active_support/cache.rb

```diff
@@ -786,6 +829,56 @@ def initialize(value, compress: true, compress_threshold: DEFAULT_COMPRESS_LIMIT
         compress!(compress_threshold) if compress
       end
 
+      def init_with(value:, compressed:, version:, expires_at:)
+        @value = value
+        @version = version
+        @created_at = 0.0
+        @expires_in = expires_at
+        @compressed = true if compressed
+        self
+      end
+
+      HEADER_FORMAT = "CGN"
+      HEADER_SIZE = [0, 0.0, 0].pack(HEADER_FORMAT).bytesize # 13 bytes
```

Comment:
The largest part of this header is the float. If `expired_in` isn't set, we're using 8 bytes to store 0.0 and then throwing it away later in `unpack`. What do you think about using another bit in the `compressed` byte to store whether `expired_in` is non-nil, and then storing the float value as an optional trailing field like `version`?

If you're really up for roflscaling this, we could pack both of those flags as the highest bits in the version length field and still have room to store version identifiers up to 1GB in size.

The pack / unpack code would get a lot gnarlier, so feel free to push back on both of these suggestions. Knocking another 9 bytes off the minimum entry size would be pretty sweet though 😁

Reply:
Yeah I thought about using that first byte as a bitfield, but I didn't think about using one of the bit for make the expiry optional.

But I think we should keep a couple bits in there in case the `Entry` format need to evolved in the future, and more importantly we need to make sure we can't match the Marshal prefix (until the legacy format would be deprecated).

So I can try that and see how much evolution it leaves us.